### DISCUSSION TITLE
Example config cannot be parsed

### DISCUSSION THREAD

**Discussion by rEnr3n** [pimalaya/himalaya#366] [Q&A] [answered]
I tested the config from wiki but I get this error:

```
$ himalaya -l trace -c test.toml
Error: cannot parse config file

Caused by:
    data did not match any variant of untagged enum DeserializedAccountConfig at line 56 column 1
```

The line in question is `[posteo-local.mailboxes]`. Anyone else got the same error?

**Comment by soywod** [comment]
Can you share your full configuration? Something must miss.

**Reply by rEnr3n** [reply]
I just copied the entire config from wiki.

```
name = "Paul DURANT"
downloads-dir = "~/Downloads"
signature = "Cordialement,\nPaul DURANT"
default-page-size = 50
watch-cmds = ["mbsync -a"]

[gmail]
default = true
email = "p.durant@gmail.com"

smtp-host = "smtp.gmail.com"
smtp-port = 465
smtp-login = "p.durant@gmail.com"
smtp-passwd-cmd = "pass show gmail"

imap-host = "imap.gmail.com"
imap-port = 993
imap-login = "p.durant@gmail.com"
imap-passwd-cmd = "pass show gmail"

[gmail.mailboxes]
sent = "[Gmail]/Sent Mail"
draft = "[Gmail]/Drafts"

[outlook]
default-page-size = 25
signature = """
Regards,
Paul DURANT
Sent from Outlook
"""

smtp-host = "smtp.office365.com"
smtp-port = 587
smtp-starttls = true
smtp-login = "pdurant@outlook.com"
smtp-passwd-cmd = "security find-internet-password -gs outlook -w"

imap-host = "outlook.office365.com"
imap-port = 993
imap-login = "pdurant@outlook.com"
imap-passwd-cmd = "security find-internet-password -gs outlook -w"

[posteo-local]
name = "Paul D."
downloads-dir = "~/Downloads/posteo"
email = "paul.durant@localhost"

smtp-host = "posteo.de"
smtp-port = 465
smtp-login = "paul.durant@posteo.com"
smtp-passwd-cmd = "security find-internet-password -gs posteo -w"

maildir-dir = "~/Maildir"

[posteo-local.mailboxes]
sent = "Sent" # will point to ~/Maildir/.Sent
draft = "~/Maildir/custom/drafts"
```

**Comment by soywod** [comment, marked as answer]
It is because the `[outlook]` section does not contain an `email` (it is
a mandatory field). I will update the wiki. Thank you for pointing it
out!

**Reply by rEnr3n** [reply]
It would be helpful if the required fields are marked as such.

**Reply by soywod** [reply]
Indeed, it is a very good remark! I invite you to update the wiki directly.

**Reply by rEnr3n** [reply]
I just started using it 🤷🏼‍♂️. You'll have to tell me which ones are. I quickly tested it. Let me know if I got it right.

Common required:
`name`
`default`
`email`
`smtp-host`
`smtp-port`
`smtp-login`
`smtp-passwd-cmd`

Required for imap only:
`imap-host`
`imap-port`
`imap-login`
`imap-passwd-cmd`

Required for maildir only:
`maildir-dir`

Required for notmuch only
`notmuch-database-dir`

**Comment by soywod** [comment]
> Common required:
> `name`
> `default`
> `email`
> `smtp-host`
> `smtp-port`
> `smtp-login`
> `smtp-passwd-cmd`

The only global required field is name. Required common (shared) fields
are email, smtp-host, smtp-port, smtp-login and smtp-passwd-cmd.

> Required for imap only:
> `imap-host`
> `imap-port`
> `imap-login`
> `imap-passwd-cmd`
> 
> Required for maildir only:
> `maildir-dir`
> 
> Required for notmuch only
> `notmuch-database-dir`

All this is correct :)

**Reply by rEnr3n** [reply]
`default` is actually needed. It gave me an error when I removed it.

**Comment by soywod** [comment]
> `default` is actually needed. It gave me an error when I removed it.

Can you share the config? In the code the default field is marked as
optional, it should not give an error.

**Reply by rEnr3n** [reply]
```
name = "A"

[sample]
email = "asd"
imap-host = "asd"
imap-port = 12
imap-login = "asd"
imap-passwd-cmd = "asd"

smtp-host = "smtp.gmail.com"
smtp-port = 465
smtp-login = "p.durant@gmail.com"
smtp-passwd-cmd = "pass show gmail"

[sample2]
email = "asd"
maildir-dir = "asd"

smtp-host = "smtp.gmail.com"
smtp-port = 465
smtp-login = "p.durant@gmail.com"
smtp-passwd-cmd = "pass show gmail"

[sample3]
email = "asd"
notmuch-database-dir = "asd"

smtp-host = "smtp.gmail.com"
smtp-port = 465
smtp-login = "p.durant@gmail.com"
smtp-passwd-cmd = "pass show gmail"
```

**Comment by soywod** [comment]
I think it is because of "sample3": to have access to
"notmuch-database-dir" you need to compile himalaya with the
"notmuch-backend" feature. Remove it and it should work without
"default".

**Reply by rEnr3n** [reply]
Removed it. I still got the same error.

```
name = "A"

[sample]
email = "asd"
imap-host = "asd"
imap-port = 12
imap-login = "asd"
imap-passwd-cmd = "asd"

smtp-host = "smtp.gmail.com"
smtp-port = 465
smtp-login = "p.durant@gmail.com"
smtp-passwd-cmd = "pass show gmail"

[sample2]
email = "asd"
maildir-dir = "asd"

smtp-host = "smtp.gmail.com"
smtp-port = 465
smtp-login = "p.durant@gmail.com"
smtp-passwd-cmd = "pass show gmail"
```

**Comment by soywod** [comment]
> Removed it. I still got the same error.

I do not have the error anymore. Which version do you use? Note that I
have the "Error: cannot find default account" because you need at least
one default account. Try with "himalaya -a sample" it should work.

**Reply by rEnr3n** [reply]
```
$ himalaya --version
himalaya 0.5.10
```

I had to set `default` when I just run `himalaya -c config.toml`. I could run `himalaya -a sample -c config.toml` when there is no `default`.

**Comment by soywod** [comment]
> I had to set `default` when I just run `himalaya -c config.toml`. I
> could run `himalaya -a sample -c config.toml` when there is no
> `default`.

Which is the normal behaviour. Otherwise Himalaya cannot know which
account to take by default.

**Comment by soywod** [comment]
> It would be helpful if the required fields are marked as such.

Yes, it is a very good remark! I invite you to update the wiki directly.

**Comment by valeratrades** [comment]
Hitting same problem with copying the now default config from <https://github.com/pimalaya/himalaya/blob/master/config.sample.toml>, - doesn't compile, throws random errors like
```
Error:
   0: cannot parse config file at "/home/v/.config/himalaya/config.toml"
   1: TOML parse error at line 236, column 1
   1:     |
   1: 236 | backend.type = "imap"
   1:     | ^^^^^^^
   1: wanted exactly 1 element, more than 1 element
   1:
```
which I couldn't get rid of by bruteforcing. Would really appreciate someone sharing their current config (I'll figure it out from there, just need a compiling baseline for reference).

// Here is my current [config that errors](<https://github.com/valeratrades/nix/blob/master/home/config/himalaya/config.toml>), but I expect it to be of little interest, as I just copied the example one and removed a bunch of things that seemed irrelevant after the original failed to compile.

P.S.: as a suggestion of how experience of both reading and maintaining it could be improved: get rid of half the comments. Many of them repeat things previously said, add extra empty indentation levels, mirror the name of the method.

**Reply by soywod** [reply]
I checked your config and you seem to use Nix with Home Manager. So I believe you are not using the `v1.0.0` because it's on the way (https://github.com/NixOS/nixpkgs/pull/363649, https://github.com/nix-community/home-manager/pull/5297). Hence the error: the config is made to work on `master`, which targets the `v1.0.0`. Error is not explicit at all due to `serde` limitations, it has been improved in the `v1.0.0`, unfortunately it's not backward-compatible. Using a recent config with old code will always result into this kind of horrible error message.

I would suggest to either wait for the merge, or use directly the branches locally.

> P.S.: as a suggestion of how experience of both reading and maintaining it could be improved: get rid of half the comments. Many of them repeat things previously said, add extra empty indentation levels, mirror the name of the method.

Thanks for the feedback, much appreciated. I will improve the sample for the next release. Feel free to open an issue with suggestions or even a PR.
